ansible_rulebook/engine.py

```diff
@@ -234,56 +233,62 @@ def __init__(
     ):
         self.pa_runner = PlaybookActionRunner()
         self.pa_runner_task = None
-        self.action_tasks = []
+        self.action_loop_task = None
+        self.action_tasks = {}
         self.event_log = event_log
         self.ruleset_queue_plan = ruleset_queue_plan
         self.name = ruleset_queue_plan.ruleset.name
         self.hosts_facts = hosts_facts
         self.variables = variables
         self.project_data_file = project_data_file
+        self.stopped = True
 
     async def run_ruleset(self):
+        self.stopped = False
         prime_facts(self.name, self.hosts_facts, self.variables)
         self.pa_runner_task = asyncio.create_task(
             self.pa_runner.start(self.name), name=self.name
         )
+        self.action_loop_task = asyncio.create_task(
+            self._drain_actionplan_queue()
+        )
+        source_loop_task = asyncio.create_task(self._drain_source_queue())
+        await asyncio.wait([source_loop_task])
 
-        logger.info("Waiting for event from %s", self.name)
+    async def _stop(self):
+        # Wait for items in queues to be completed. Mainly for spec tests.
+        await asyncio.sleep(0.01)
+
+        logger.info("Attempt to stop ruleset %s", self.name)
+        self.stopped = True
+        self.pa_runner.stop()
+        # helps to break waiting on the plan queue when it is empty
+        try:
+            self.ruleset_queue_plan.plan.queue.put_nowait("STOP")
```

Comment:
@bzwei Why do we have to put this in the plan queue, which typically contains ActionContext objects. You could have multiple ActionContexts queued but when a shutdown comes the queue would contain
```
    ["ActionContext1","ActionContext2","ActionContext3","STOP"]
```
So this would wait till all ActionContexts are drained before it gets to the STOP

We should be able to stop the action_plan task when the source queue gets a Shutdown

Reply:
If we ran 2 tasks, the source_queue_task and action_plan_queue_task and run them separately, store each one in the ruleset_tasks. When the source_queue_task gets a "Shutdown" it ends the ruleset and returns when a task is completed we cancel all the ruleset_tasks, which will stop the action_plan_queue_task. It doesn't need special processing

Reply:
Let me try it out. Thanks